Increasing test coverage of rclcpp_action

brawner opened the pull request:
As part of the effort to bring the rclcpp packages up to Quality Level 1, this introduces many more tests of the rclcpp_action api. This brings the coverage from 76% to 83% of the `src` files and 94% of the `include` files (87% combined). 

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=9862)](http://ci.ros2.org/job/ci_linux/9862/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=5478)](http://ci.ros2.org/job/ci_linux-aarch64/5478/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=8022)](http://ci.ros2.org/job/ci_osx/8022/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=9718)](http://ci.ros2.org/job/ci_windows/9718/)
* Coverage [![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_linux_coverage&build=78)](https://ci.ros2.org/job/ci_linux_coverage/78/)

Signed-off-by: Stephen Brawner <brawner@gmail.com>

ahcorde commented on rclcpp_action/test/test_server.cpp:
```diff
@@ -95,6 +95,32 @@ TEST_F(TestServer, construction_and_destruction)
   (void)as;
 }
 
+TEST_F(TestServer, construction_and_destruction_callback_group)
```
There is no `EXPECT_*` or `ASSERT_*` in this test. What are you trying to test?

brawner replied:
Like the construction_and_destruction test above, it just simply checks that creating and cleaning up the object doesn't throw exceptions or cause segfaults. I didn't really have any other ideas for how to add coverage for the callback group constructor other than this simple test.

I'm open to ideas for better exercising that constructor though.

ahcorde replied:
you can do something like:

```
  try{
     <test code>
  }
  catch (Exception& e)
  {
    FAIL();
  }

  SUCCEED();

brawner replied:
Added an ASSERT_NO_THROW around this code, which does the same thing.

brawner commented:
@ahcorde ping for review

brawner commented:
Rebased and addressed feedback @ahcorde 

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=10365)](http://ci.ros2.org/job/ci_linux/10365/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=5881)](http://ci.ros2.org/job/ci_linux-aarch64/5881/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=8441)](http://ci.ros2.org/job/ci_osx/8441/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=10262)](http://ci.ros2.org/job/ci_windows/10262/)

ahcorde commented on rclcpp_action/test/test_types.cpp:
```diff
@@ -0,0 +1,62 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
```
Remove one line

brawner replied:
Done

ahcorde commented:
In windows CI test_types is passing but in the result I see this error:

```
02:02:30 [MSBuild] [-ERROR-] Can't create fingerprints for some files:
02:02:30 [MSBuild] [-ERROR-] - 'C:/ci/ws/src/ros2/rclcpp/rclcpp_action/test/test_types.cpp', IO exception has been thrown: java.nio.file.NoSuchFileException: C:\ci\ws\src\ros2\rclcpp\rclcpp_action\test\test_types.cpp
02:02:30 [MSBuild] [-ERROR-] - 'C:/ci/ws/src/ros2/rclcpp/rclcpp_action/test/test_types.cpp', IO exception has been thrown: java.nio.file.NoSuchFileException: C:\ci\ws\src\ros2\rclcpp\rclcpp_action\test\test_types.cpp
02:02:30 [MSBuild] [-ERROR-] - 'C:/ci/ws/src/ros2/rclcpp/rclcpp_action/test/test_types.cpp', IO exception has been thrown: java.nio.file.NoSuchFileException: C:\ci\ws\src\ros2\rclcpp\rclcpp_action\test\test_types.cpp
02:02:30 [MSBuild] [-ERROR-] - 'C:/ci/ws/src/ros2/rclcpp/rclcpp_action/test/test_types.cpp', IO exception has been thrown: java.nio.file.NoSuchFileException: C:\ci\ws\src\ros2\rclcpp\rclcpp_action\test\test_types.cpp
02:02:30 [MSBuild] [-ERROR-] - 'C:/ci/ws/src/ros2/rclcpp/rclcpp_action/test/test_types.cpp', IO exception has been thrown: java.nio.file.NoSuchFileException: C:\ci\ws\src\ros2\rclcpp\rclcpp_action\test\test_types.cpp
```

brawner commented:
I'm not sure what the error that fingerprinting error relates to, but the ci_windows build was marked because I used the wrong unsigned type in the test_types.cpp file. I've fixed that, running the tests again.

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=10409)](http://ci.ros2.org/job/ci_linux/10409/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=5917)](http://ci.ros2.org/job/ci_linux-aarch64/5917/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=8489)](http://ci.ros2.org/job/ci_osx/8489/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=10299)](http://ci.ros2.org/job/ci_windows/10299/)

ahcorde commented:
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=8571)](http://ci.ros2.org/job/ci_osx/8571/)

brawner commented:
Rebasing this branch and my ros2.repos file.

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=10540)](http://ci.ros2.org/job/ci_linux/10540/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=5996)](http://ci.ros2.org/job/ci_linux-aarch64/5996/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=8586)](http://ci.ros2.org/job/ci_osx/8586/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=10412)](http://ci.ros2.org/job/ci_windows/10412/)

ahcorde commented:
@brawner 
 - two related warrning on macOS
     - https://ci.ros2.org/job/ci_osx/8586/clang/new/folder.1894854374/fileName.-821771868/source.0e87ab63-24c4-48ce-bad9-5c4a29e909d6/#276
   - https://ci.ros2.org/job/ci_osx/8586/clang/new/folder.1894854374/fileName.544833820/source.dfc01374-09e5-4b83-a612-a14192280bd1/#102
